### PR TITLE
Improve docstring of Axes.pcolor

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

As part of #10148: Docstring update for `Axes.pcolor`.

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -5338,39 +5345,46 @@ def pcolor(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
...
         --------
         pcolormesh : for an explanation of the differences between
             pcolor and pcolormesh.
+        imshow : If *X* and *Y* are each equidistant, `~.Axes.imshow` can be a
+            faster alternative.
 
         Notes
         -----
+
+        **Masked arrays**
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: does numpydoc not support lower-level headings?
```
Masked arrays
~~~~~~~~~~~~~
```
(dunno)

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -5338,39 +5345,46 @@ def pcolor(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
...
+
+        *X*, *Y* and *C* may be masked arrays. If either ``C[i, j]``, or one
+        of the vertices surrounding ``C[i,j]`` (*X* or *Y* at
+        ``[i, j], [i+1, j], [i, j+1], [i+1, j+1]``) is masked, nothing is
+        plotted.
+
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: as there's no heading there I'm not sure this anchor is correct (see above re lower headings)

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -5338,39 +5345,46 @@ def pcolor(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
         Other Parameters
         ----------------
         antialiaseds : bool, optional, default: False
-            The default ``antialiaseds`` is False if the default
-            ``edgecolors="none"`` is used.  This eliminates artificial lines
+            The default *antialiaseds* is False if the default
+            *edgecolors*\ ="none" is used.  This eliminates artificial lines
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: space seems extraneous?

File: lib/matplotlib/axes/_axes.py [hunk 4]
```diff
@@ -5268,62 +5268,69 @@ def _pcolorargs(funcname, *args, allmatch=False):
     def pcolor(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
                vmax=None, **kwargs):
         """
-        Create a pseudocolor plot of a 2-D array.
+        Create a pseudocolor plot of a 2-D array using quadrilaterals.
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: Not convinced "using quadrilaterals" really helps here (and yes I know how pcolor is implemented).

File: lib/matplotlib/axes/_axes.py (hunk 1 above)
Comment (timhoffm) [reply lib/matplotlib/axes/_axes.py]: Nope. If you try this, numpydoc simply swallows the lower-level headings.

I think structuring with bold paragraphs is the best one can do.

File: lib/matplotlib/axes/_axes.py (hunk 2 above)
Comment (timhoffm) [reply lib/matplotlib/axes/_axes.py]: This does work the id is attached to the paragraph:

`<p id="axes-pcolor-grid-orientation"><strong>Grid orientation</strong></p>`

File: lib/matplotlib/axes/_axes.py (hunk 4 above)
Comment (timhoffm) [reply lib/matplotlib/axes/_axes.py]: I think we need some sort of addition to differentiate it from `imshow` and `matshow`. A better wording is welcome.

Comment (jklymak) [reply lib/matplotlib/axes/_axes.py]: Its explained below - I think the differentiation is useful.  OTOH to me the main difference w/ `imshow` is that the lateral and vertical spacing can be non-even (as opposed to distorted grids.)

File: lib/matplotlib/axes/_axes.py [hunk 5]
```diff
@@ -5268,62 +5268,69 @@ def _pcolorargs(funcname, *args, allmatch=False):
...
+        Call signature::
 
-            pcolor(C, **kwargs)
-            pcolor(X, Y, C, **kwargs)
+            pcolor([X, Y,] C, **kwargs)
 
-        pcolor can be very slow for large arrays; consider
-        using the similar but much faster
-        :func:`~matplotlib.pyplot.pcolormesh` instead.
+        *X* and *Y* can be used to specify the corners of the quadrilatals.
```
Comment (jklymak) [inline lib/matplotlib/axes/_axes.py]: Typo: quadrilaterals

File: lib/matplotlib/axes/_axes.py [hunk 6]
```diff
@@ -5338,39 +5345,46 @@ def pcolor(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
         Other Parameters
         ----------------
         antialiaseds : bool, optional, default: False
```
Comment (jklymak) [inline lib/matplotlib/axes/_axes.py]: wow, is this really the parameter name?  Thats terrible!

File: lib/matplotlib/axes/_axes.py [hunk 7]
```diff
@@ -5338,39 +5345,46 @@ def pcolor(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
...
-        of the vertices surrounding C[i,j] (``X`` or ``Y`` at [i, j], [i+1, j],
-        [i, j+1], [i+1, j+1]) is masked, nothing is plotted.
+        **Grid orientation**
 
-        The grid orientation follows the MATLAB convention: an array ``C`` with
-        shape (nrows, ncolumns) is plotted with the column number as ``X`` and
-        the row number as ``Y``, increasing up; hence it is plotted the way the
-        array would be printed, except that the ``Y`` axis is reversed. That
-        is, ``C`` is taken as ``C`` (y, x).
+        The grid orientation follows the MATLAB convention: an array *C* with
```
Comment (jklymak) [inline lib/matplotlib/axes/_axes.py]: I don't particularly like this note, and find it confusing as written. Can we improve it while here?  

```python
"""
An array *C* with shape (nrows, columns) is plotted with the column number *X* 
and the row number as *Y*.  
"""
```
suffices in my opinion.  I don't understand why anyone would think that `y[j]` doesn't correspond to `C[j, i]` so the only ambiguous orientation is if `C` is organized as `C[j, i]` or `C[i, j]`.

File: lib/matplotlib/axes/_axes.py [hunk 8]
```diff
@@ -5400,10 +5414,12 @@ def pcolor(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
 
             pcolor(C.T)
 
-        MATLAB :func:`pcolor` always discards the last row and column of ``C``,
-        but Matplotlib displays the last row and column if ``X`` and ``Y`` are
-        not specified, or if ``X`` and ``Y`` have one more row and column than
-        ``C``.
+        **Difference to MATLAB pcolor()**
```
Comment (jklymak) [inline lib/matplotlib/axes/_axes.py]: again:

**Handling of pcolor end-cases**

`pcolor` displays all columns of *C* if *X* and *Y* are not specified, or if *X* and *Y* have one more column than *C*.  If *X* and *Y* have the same number of columns as *C* then the last column of *C* is dropped.   Similarly for the rows.

File: lib/matplotlib/axes/_axes.py (hunk 4 above)
Comment (timhoffm) [reply lib/matplotlib/axes/_axes.py]: I could also do "Create a pseudocolor plot with a non-regular grid."

File: lib/matplotlib/axes/_axes.py (hunk 6 above)
Comment (timhoffm) [reply lib/matplotlib/axes/_axes.py]: Yes, we have this pluralization scheme in some places, see e.g. LineCollection: colors, edgecolors, antialiaseds, ... Not really great and IMO not really necessary to pluralize, but it's the way it is now.

File: lib/matplotlib/axes/_axes.py (hunk 7 above)
Comment (timhoffm) [reply lib/matplotlib/axes/_axes.py]: Significantly shortend to:

The grid orientation follows the standard matrix convention: An array
*C* with shape (nrows, ncolumns) is plotted with the column number as
*X* and the row number as *Y*.

Removed also the whole example stuff as it's more confusing than helpflul.

File: lib/matplotlib/axes/_axes.py (hunk 4 above)
Comment (jklymak) [reply lib/matplotlib/axes/_axes.py]: I like that wording!

**Comment by timhoffm** [comment]
@meeseeksdev backport to v2.2.x

**Comment by jklymak** [comment]
...we aren't backporting *.py files...

**Comment by timhoffm** [comment]
That does only hold for the branch v2.2-doc. I'm quite strongly for backporting pure docstring updates to 2.2.x. It will be around for quite some time and definitively deserves the docstring improvements.

**Comment by jklymak** [comment]
I'm not at all the final say on these things, so please feel free to argue up the chain... But I think 2.2.x is just for regressions and other important bug fixes.  I don't think we want to also be merging in a bunch of doc string changes.  I agree they are hugely helpful, but folks have managed to muddle through with the old docs so I personally don't think its critical that they be backported.  But I don't actually do the backporting...

File: lib/matplotlib/axes/_axes.py [hunk 9]
```diff
@@ -5338,72 +5349,54 @@ def pcolor(self, *args, alpha=None, norm=None, cmap=None, vmin=None,
         Other Parameters
         ----------------
         antialiaseds : bool, optional, default: False
-            The default ``antialiaseds`` is False if the default
-            ``edgecolors="none"`` is used.  This eliminates artificial lines
+            The default *antialiaseds* is False if the default
+            *edgecolors*\ ="none" is used.  This eliminates artificial lines
```
Comment (jklymak) [inline lib/matplotlib/axes/_axes.py]: I can't run the tests wth this line:  `SyntaxError: invalid escape sequence \`

Comment (timhoffm) [reply lib/matplotlib/axes/_axes.py]: Strange. This is the [recommended way](http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#inline-markup):

     Use a backslash escaped space to work around that: thisis\ *one*\ word.

The docs itself look fine:
https://matplotlib.org/devdocs/api/_as_gen/matplotlib.axes.Axes.pcolor.html?highlight=pcolor#matplotlib.axes.Axes.pcolor

Why is the test inspecting the docstring at all? Do we have some doctests somewhere that need to be run?

Comment (QuLogic) [reply lib/matplotlib/axes/_axes.py]: @timhoffm The recommended way you cite is for Sphinx, while @jklymak's error is coming from Python. The backslash escape for Sphinx is on the text that it sees, which is either a reST file or a docstring (extracted to a reST file by auto-something), but in order to place the backslash in the docstring, you _also_ need to escape it for Python. So you need to double-escape or make it a raw string.

As to why @jklymak only sees it on one machine, maybe it's `tests.py` vs `pytest`? The former turns this deprecation warnings into errors. Also, the warnings are only raised on Python 3.6+.

Comment (jklymak) [reply lib/matplotlib/axes/_axes.py]: Ahh. Thanks @QuLogic :  `python tests.py` does not work and gives the Syntax Error.  `py.test` works.  Thats what I get for just hitting `^R` for the last time I ran the tests - obviously on one machine I mentally upgraded to `py.test` and the other I haven't :wink:...

**Pull request merged**